File: NS_fluids_lib/LSA_Rayleigh_Python/tests/test_end_to_end.py
"""End-to-end test: real AMReX plotfiles -> twin subtraction -> DMD -> sigma.

Synthetic plotfiles are written in the genuine on-disk AMReX format and read back through the production code path (nsmfp.read_plotfile -> yt -> covering_grid).  Nothing here is mocked, so a format or ordering mistake in the reader shows up as a wrong growth rate rather than passing silently.

The planted signal mimics the real experiment:

    Q_unperturbed(t) = drift(t)                      (parasitic currents)
    Q_perturbed(t)   = drift(t) + eps e^{sigma t} phi(z)

so the twin subtraction must remove `drift` exactly and leave a clean exponentially growing capillary mode whose rate DMD recovers.  `drift` is made deliberately large compared with the perturbation - as spurious currents often are in a surface-tension calculation - so the test fails if the subtraction is skipped or mispaired.
"""

import os
import sys

import numpy as np
import pytest

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "python"))
sys.path.insert(0, os.path.dirname(__file__))

import dispersion  # noqa: E402
import dmd as dmd_mod  # noqa: E402
import nsmfp  # noqa: E402
import synthetic_plotfile as sp  # noqa: E402

yt = pytest.importorskip("yt")
yt.set_log_level(50)

NX, NY = 10, 32
KR0 = 0.7
R0 = 1.0
LAMBDA = 2.0 * np.pi / (KR0 / R0)
SIGMA_EXACT = float(dispersion.growth_rate(KR0 / R0, r0=R0))
EPS = 1e-3
DT = 0.05
NSNAP = 60


def _z(ny=NY):
    return (np.arange(ny) + 0.5) * LAMBDA / ny


def _mode_shape(nx=NX, ny=NY):
    """Interface-localised sinusoid: cos(k z), tapered in r."""
    z = _z(ny)
    r = (np.arange(nx) + 0.5) * (4.0 * R0) / nx
    taper = np.exp(-((r - R0) ** 2) / 0.25)
    return np.outer(taper, np.cos(KR0 * z / R0))


def _drift(t, nx=NX, ny=NY):
    """Base-state drift (parasitic currents): large, slow, non-modal."""
    z = _z(ny)
    r = (np.arange(nx) + 0.5) * (4.0 * R0) / nx
    return 0.5 * np.outer(np.sin(3.0 * r), np.cos(2.0 * z)) * (1.0 + 0.1 * t)


@pytest.fixture(scope="module")
def twin_runs(tmp_path_factory):
    """Write an unperturbed twin and a perturbed run as real plotfiles."""
    root = tmp_path_factory.mktemp("runs")
    base_dir = str(root / "base")
    pert_dir = str(root / "pert")

    steps = list(range(0, NSNAP * 10, 10))
    times = [i * DT for i in range(NSNAP)]
    shape_x = _mode_shape()
    shape_y = 0.6 * _mode_shape()

    def base_fn(t, name, nx, ny):
        return _drift(t, nx, ny)

    def pert_fn(t, name, nx, ny):
        growth = EPS * np.exp(SIGMA_EXACT * t)
        shape = shape_x if name == "x_velocity" else shape_y
        return _drift(t, nx, ny) + growth * shape

    sp.write_series(base_dir, base_fn, steps, times, shape=(NX, NY),
                    domain_right=(4.0 * R0, LAMBDA))
    sp.write_series(pert_dir, pert_fn, steps, times, shape=(NX, NY),
                    domain_right=(4.0 * R0, LAMBDA))
    return pert_dir, base_dir


def test_find_plotfiles_sorts_numerically(twin_runs):
    pert, _ = twin_runs
    paths = nsmfp.find_plotfiles(pert)
    assert len(paths) == NSNAP
    steps = [nsmfp.plotfile_step(p) for p in paths]
    assert steps == sorted(steps)


def test_read_plotfile_shape_and_time(twin_runs):
    pert, _ = twin_runs
    pd = nsmfp.read_plotfile(nsmfp.find_plotfiles(pert)[3])
    assert pd.data.size == NX * NY * 2          # two fields concatenated
    assert pd.time == pytest.approx(3 * DT)
    assert pd.step == 30


def test_twin_subtraction_removes_the_drift(twin_runs):
    """|Q'| must reflect only the planted perturbation, not the drift."""
    pert, base = twin_runs
    X, dt, times = nsmfp.build_snapshot_matrix(pert, base, verbose=False)
    assert dt == pytest.approx(DT)
    assert X.shape[1] == NSNAP

    expected0 = EPS * np.linalg.norm(
        np.concatenate([_mode_shape().ravel(), 0.6 * _mode_shape().ravel()]))
    assert np.linalg.norm(X[:, 0]) == pytest.approx(expected0, rel=1e-10)

    # the drift itself is far larger than the perturbation, so this is a meaningful check that it really cancelled
    drift_norm = np.linalg.norm(_drift(0.0).ravel())
    assert drift_norm > 100 * np.linalg.norm(X[:, 0])


def test_end_to_end_growth_rate_matches_dispersion_relation(twin_runs):
    """The headline check: recovered sigma equals the exact Rayleigh value."""
    pert, base = twin_runs
    X, dt, _ = nsmfp.build_snapshot_matrix(pert, base, verbose=False)
    res = dmd_mod.dmd(X, dt)
    measured = res.leading_growth_rate()
    assert measured == pytest.approx(SIGMA_EXACT, rel=1e-6)
    assert 0.3 < measured < 0.4          # the physically expected magnitude


def test_end_to_end_norm_fit_agrees_with_dmd(twin_runs):
    pert, base = twin_runs
    times, norms = nsmfp.perturbation_norms(pert, base)
    rate, _, r2 = nsmfp.fit_exponential_growth(times, norms)
    assert rate == pytest.approx(SIGMA_EXACT, rel=1e-8)
    assert r2 > 0.999999


def test_mispaired_steps_are_rejected(twin_runs, tmp_path):
    """Deleting a twin plotfile must not silently shift the pairing."""
    pert, base = twin_runs
    import shutil

    trimmed = str(tmp_path / "base_trimmed")
    shutil.copytree(base, trimmed)
    # remove one interior snapshot from the twin
    victim = nsmfp.find_plotfiles(trimmed)[5]
    shutil.rmtree(victim)

    # Pairing is by step number, so the survivors still align; the missing step is dropped, which makes the sampling non-uniform and must raise.
    with pytest.raises(ValueError, match="not uniform"):
        nsmfp.build_snapshot_matrix(pert, trimmed, verbose=False)


def test_grid_mismatch_is_rejected(tmp_path):
    """Twins on different grids must not be subtracted."""
    a = str(tmp_path / "a")
    b = str(tmp_path / "b")
    sp.write_series(a, lambda t, n, nx, ny: np.ones((nx, ny)),
                    [0, 10], [0.0, DT], shape=(8, 16))
    sp.write_series(b, lambda t, n, nx, ny: np.ones((nx, ny)),
                    [0, 10], [0.0, DT], shape=(8, 32))
    with pytest.raises(ValueError, match="grid mismatch"):
        nsmfp.build_snapshot_matrix(a, b, verbose=False)


def test_time_mismatch_is_rejected(tmp_path):
    """Same steps but different times means the decks disagreed on dt."""
    a = str(tmp_path / "a")
    b = str(tmp_path / "b")
    sp.write_series(a, lambda t, n, nx, ny: np.ones((nx, ny)),
                    [0, 10], [0.0, 0.05], shape=(8, 16))
    sp.write_series(b, lambda t, n, nx, ny: np.ones((nx, ny)),
                    [0, 10], [0.0, 0.07], shape=(8, 16))
    with pytest.raises(ValueError, match="time mismatch"):
        nsmfp.build_snapshot_matrix(a, b, verbose=False)


def test_missing_directory_raises(tmp_path):
    with pytest.raises(FileNotFoundError):
        nsmfp.build_snapshot_matrix(str(tmp_path / "nope"),
                                    str(tmp_path / "also_nope"), verbose=False)


def test_skip_and_stride(twin_runs):
    """Striding must scale dt, and still give the same growth rate."""
    pert, base = twin_runs
    X, dt, _ = nsmfp.build_snapshot_matrix(pert, base, skip=10, stride=3,
                                           verbose=False)
    assert dt == pytest.approx(3 * DT)
    assert X.shape[1] == len(range(10, NSNAP, 3))
    assert dmd_mod.dmd(X, dt).leading_growth_rate() == pytest.approx(
        SIGMA_EXACT, rel=1e-6)


def test_single_field_subspace_still_works(twin_runs):
    """Ranjan et al. note one well-chosen variable can suffice (and cuts
    memory up to 80%).  Check the pipeline honours a single-field subspace."""
    pert, base = twin_runs
    X, dt, _ = nsmfp.build_snapshot_matrix(pert, base, fields=("x_velocity",),
                                           verbose=False)
    assert X.shape[0] == NX * NY
    assert dmd_mod.dmd(X, dt).leading_growth_rate() == pytest.approx(
        SIGMA_EXACT, rel=1e-6)


def test_identical_runs_give_zero_perturbation(twin_runs):
    """Subtracting a run from itself must give exactly zero - the guard the
    driver uses to detect a perturbed deck that forgot to set radblob."""
    _, base = twin_runs
    X, _, _ = nsmfp.build_snapshot_matrix(base, base, verbose=False)
    assert np.allclose(X, 0.0)

File: NS_fluids_lib/LSA_Rayleigh_Fortran/tests/test_end_to_end.f90
!> End-to-end test: real AMReX plotfiles -> twin subtraction -> DMD -> sigma.
!>
!> Plotfiles are written in the genuine on-disk AMReX format and read back through the production path (plotfile_mod), so a format or ordering mistake in the hand-written parser shows up as a wrong growth rate rather than passing silently.  The same synthetic files are readable by yt, which is how the format was cross-checked independently of this code.
!>
!> The planted signal mimics the real experiment:
!>
!> Q_unperturbed(t) = drift(t)                       (parasitic currents) Q_perturbed(t)   = drift(t) + eps e^{sigma t} phi(z)
!>
!> so the twin subtraction must remove `drift` exactly and leave a clean exponentially growing capillary mode.  `drift` is made 100x larger than the perturbation - as spurious currents often are in a surface-tension calculation - so the test fails if the subtraction is skipped or mispaired.
program test_end_to_end
   use testing_mod
   use plotfile_mod
   use plotfile_writer_mod
   use nsmfp_mod
   use dmd_mod
   use dispersion_mod, only: growth_rate, PI
   implicit none

   integer, parameter :: NX = 10, NY = 32, NSNAP = 60
   real(dp), parameter :: KR0 = 0.7_dp, R0 = 1.0_dp
   real(dp), parameter :: EPSA = 1.0e-3_dp, DT = 0.05_dp

   real(dp) :: lambda, sexact, lo(2), hi(2)
   character(len=64) :: names(2), want(2)
   character(len=256) :: pdir, bdir
   real(dp), allocatable :: X(:,:), times(:), norms(:)
   real(dp) :: dt_out, slope, intercept, r2
   type(dmd_result_t) :: res
   type(plotfile_t) :: pf
   integer :: ierr, i, j, k, step
   real(dp) :: shape_x(NX,NY), shape_y(NX,NY), d(NX,NY,2)
   real(dp) :: t, growth, expected0, drift_norm
   character(len=MAX_NAME), allocatable :: paths(:)
   integer :: nfound

   lambda = 2.0_dp * PI / (KR0 / R0)
   sexact = growth_rate(KR0/R0, R0, 1.0_dp, 1.0_dp)
   names(1) = 'x_velocity'; names(2) = 'y_velocity'
   want(1) = 'x_velocity';  want(2) = 'y_velocity'
   lo = [0.0_dp, 0.0_dp]
   hi = [4.0_dp*R0, lambda]
   pdir = '/tmp/lsa_e2e/pert'
   bdir = '/tmp/lsa_e2e/base'

   call execute_command_line('rm -rf /tmp/lsa_e2e')
   call execute_command_line('mkdir -p '//trim(pdir)//' '//trim(bdir))

   call build_shapes()
   call write_runs()

   call section('plotfile discovery')
   call find_plotfiles(trim(pdir), paths, nfound)
   call check_int(nfound, NSNAP, 'finds every perturbed plotfile')
   call check(plotfile_step(paths(1)) < plotfile_step(paths(nfound)), &
              'plotfiles are sorted by step number')
   call check_int(plotfile_step('/x/plt00120'), 120, 'parses a step number')
   call check_int(plotfile_step('/x/plt00009'), 9, 'parses a padded step')

   call section('single plotfile read')
   call read_plotfile(trim(paths(4)), pf, ierr, want)
   call check_int(ierr, 0, 'read succeeds')
   call check_int(pf%nx, NX, 'nx correct')
   call check_int(pf%ny, NY, 'ny correct')
   call check_int(pf%ncomp, 2, 'component count correct')
   call check_close(pf%time, 3.0_dp*DT, 1.0e-12_dp, 'time correct')
   call check_int(pf%step, 30, 'step correct')
   call free_plotfile(pf)

   call section('twin subtraction removes the drift')
   call build_snapshot_matrix(trim(pdir), trim(bdir), want, 0, 1, 0, &
                              X, times, dt_out, ierr, verbose=.false.)
   call check_int(ierr, 0, 'subspace assembly succeeds')
   ! Guard: if assembly failed, X is unallocated and every later check would dereference it.  Stop cleanly rather than segfaulting on a bad diagnosis.
   if (ierr /= 0) then
      write(*,'(a)') '  cannot continue without a subspace'
      call summary('test_end_to_end')
   end if
   call check_close(dt_out, DT, 1.0e-12_dp, 'uniform dt detected')
   call check_int(size(X,2), NSNAP, 'all snapshots retained')
   call check_int(size(X,1), NX*NY*2, 'state vector size = nx*ny*ncomp')

   expected0 = EPSA * sqrt(sum(shape_x**2) + sum(shape_y**2))
   call check_close_rel(norm2(X(:,1)), expected0, 1.0e-10_dp, &
                        'first snapshot equals the planted perturbation')
   drift_norm = sqrt(sum(drift(0.0_dp)**2))
   call check(drift_norm > 100.0_dp*norm2(X(:,1)), &
              'drift really was 100x larger than the perturbation')

   call section('growth rate from the norm fit')
   call perturbation_norms(X, norms)
   call fit_exponential_growth(times, norms, times(1), -1.0_dp, &
                               slope, intercept, r2, ierr)
   call check_int(ierr, 0, 'fit succeeds')
   call check_close_rel(slope, sexact, 1.0e-8_dp, &
                        'norm slope equals the exact Rayleigh rate')
   call check(r2 > 0.999999_dp, 'fit is essentially perfect (r^2 > 0.999999)')

   call section('growth rate from DMD (headline check)')
   call dmd_compute(X, dt_out, 0, 1.0e-10_dp, res, ierr)
   call check_int(ierr, 0, 'dmd succeeds')
   call check_close_rel(leading_growth_rate(res), sexact, 1.0e-6_dp, &
                        'DMD recovers the exact Rayleigh growth rate')
   call check(leading_growth_rate(res) > 0.3_dp .and. &
              leading_growth_rate(res) < 0.4_dp, &
              'growth rate has the expected magnitude ~0.343')
   call dmd_free(res)
   deallocate(X, times, norms)

   call section('single-field subspace')
   ! Ranjan et al. note one well-chosen variable can suffice, cutting memory up to 80%.  The pipeline must honour a single-component request.
   call build_snapshot_matrix(trim(pdir), trim(bdir), want(1:1), 0, 1, 0, &
                              X, times, dt_out, ierr, verbose=.false.)
   call check_int(ierr, 0, 'single-field assembly succeeds')
   call check_int(size(X,1), NX*NY, 'state vector holds one component')
   call dmd_compute(X, dt_out, 0, 1.0e-10_dp, res, ierr)
   call check_close_rel(leading_growth_rate(res), sexact, 1.0e-6_dp, &
                        'single field still gives the right growth rate')
   call dmd_free(res); deallocate(X, times)

   call section('skip and stride')
   call build_snapshot_matrix(trim(pdir), trim(bdir), want, 10, 3, 0, &
                              X, times, dt_out, ierr, verbose=.false.)
   call check_int(ierr, 0, 'skip/stride assembly succeeds')
   call check_close(dt_out, 3.0_dp*DT, 1.0e-12_dp, 'stride scales dt')
   call dmd_compute(X, dt_out, 0, 1.0e-10_dp, res, ierr)
   call check_close_rel(leading_growth_rate(res), sexact, 1.0e-6_dp, &
                        'strided subspace gives the same growth rate')
   call dmd_free(res); deallocate(X, times)

   call section('guard rails')
   ! identical runs must give exactly zero: the check the driver uses to detect a perturbed deck that forgot to set radblob
   call build_snapshot_matrix(trim(bdir), trim(bdir), want, 0, 1, 0, &
                              X, times, dt_out, ierr, verbose=.false.)
   call check_int(ierr, 0, 'self-subtraction assembles')
   call check_close(maxval(abs(X)), 0.0_dp, 0.0_dp, &
                    'subtracting a run from itself gives exactly zero')
   deallocate(X, times)

   call test_grid_mismatch()
   call test_time_mismatch()
   call test_missing_dir()
   call test_nonuniform()

   call summary('test_end_to_end')

contains

   subroutine build_shapes()
      integer :: i, j
      real(dp) :: z, r, taper
      do j = 1, NY
         z = (real(j,dp) - 0.5_dp) * lambda / real(NY,dp)
         do i = 1, NX
            r = (real(i,dp) - 0.5_dp) * (4.0_dp*R0) / real(NX,dp)
            taper = exp(-((r - R0)**2) / 0.25_dp)
            shape_x(i,j) = taper * cos(KR0 * z / R0)
            shape_y(i,j) = 0.6_dp * shape_x(i,j)
         end do
      end do
   end subroutine build_shapes

   !> Base-state drift (parasitic currents): large, slow, non-modal.
   function drift(t) result(a)
      real(dp), intent(in) :: t
      real(dp) :: a(NX,NY)
      integer :: i, j
      real(dp) :: z, r
      do j = 1, NY
         z = (real(j,dp) - 0.5_dp) * lambda / real(NY,dp)
         do i = 1, NX
            r = (real(i,dp) - 0.5_dp) * (4.0_dp*R0) / real(NX,dp)
            a(i,j) = 0.5_dp * sin(3.0_dp*r) * cos(2.0_dp*z) * (1.0_dp + &
               0.1_dp*t)
         end do
      end do
   end function drift

   subroutine write_runs()
      integer :: j, step
      real(dp) :: t, growth
      character(len=256) :: p
      do j = 1, NSNAP
         t = real(j-1, dp) * DT
         step = (j-1) * 10
         growth = EPSA * exp(sexact * t)

         d(:,:,1) = drift(t)
         d(:,:,2) = drift(t)
         write(p,'(a,a,i5.5)') trim(bdir), '/nddataPLT', step
         call write_plotfile(trim(p), d, names, t, step, lo, hi)

         d(:,:,1) = drift(t) + growth*shape_x
         d(:,:,2) = drift(t) + growth*shape_y
         write(p,'(a,a,i5.5)') trim(pdir), '/nddataPLT', step
         call write_plotfile(trim(p), d, names, t, step, lo, hi)
      end do
   end subroutine write_runs

   subroutine test_grid_mismatch()
      real(dp) :: a(8,16,1), b(8,32,1), l(2), h(2)
      character(len=64) :: nm(1)
      real(dp), allocatable :: XX(:,:), tt(:)
      real(dp) :: dd
      integer :: e
      nm(1) = 'x_velocity'
      l = [0.0_dp,0.0_dp]; h = [1.0_dp,1.0_dp]
      a = 1.0_dp; b = 1.0_dp
      call execute_command_line('rm -rf /tmp/lsa_gm && mkdir -p /tmp/lsa_gm/a /tmp/lsa_gm/b')
      call write_plotfile('/tmp/lsa_gm/a/nddataPLT00000', a, nm, 0.0_dp, 0, l, h)
      call write_plotfile('/tmp/lsa_gm/a/nddataPLT00010', a, nm, DT, 10, l, h)
      call write_plotfile('/tmp/lsa_gm/b/nddataPLT00000', b, nm, 0.0_dp, 0, l, h)
      call write_plotfile('/tmp/lsa_gm/b/nddataPLT00010', b, nm, DT, 10, l, h)
      call build_snapshot_matrix('/tmp/lsa_gm/a', '/tmp/lsa_gm/b', nm, &
                                 0, 1, 0, XX, tt, dd, e, verbose=.false.)
      call check_int(e, 5, 'grid mismatch is rejected')
   end subroutine test_grid_mismatch

   subroutine test_time_mismatch()
      real(dp) :: a(8,16,1), l(2), h(2)
      character(len=64) :: nm(1)
      real(dp), allocatable :: XX(:,:), tt(:)
      real(dp) :: dd
      integer :: e
      nm(1) = 'x_velocity'
      l = [0.0_dp,0.0_dp]; h = [1.0_dp,1.0_dp]
      a = 1.0_dp
      call execute_command_line('rm -rf /tmp/lsa_tm && mkdir -p /tmp/lsa_tm/a /tmp/lsa_tm/b')
      call write_plotfile('/tmp/lsa_tm/a/nddataPLT00000', a, nm, 0.0_dp, 0, l, h)
      call write_plotfile('/tmp/lsa_tm/a/nddataPLT00010', a, nm, 0.05_dp, 10, l, h)
      call write_plotfile('/tmp/lsa_tm/b/nddataPLT00000', a, nm, 0.0_dp, 0, l, h)
      call write_plotfile('/tmp/lsa_tm/b/nddataPLT00010', a, nm, 0.07_dp, 10, l, h)
      call build_snapshot_matrix('/tmp/lsa_tm/a', '/tmp/lsa_tm/b', nm, &
                                 0, 1, 0, XX, tt, dd, e, verbose=.false.)
      call check_int(e, 5, 'time mismatch is rejected')
   end subroutine test_time_mismatch

   subroutine test_missing_dir()
      character(len=64) :: nm(1)
      real(dp), allocatable :: XX(:,:), tt(:)
      real(dp) :: dd
      integer :: e
      nm(1) = 'x_velocity'
      call build_snapshot_matrix('/tmp/lsa_nope_a', '/tmp/lsa_nope_b', nm, &
                                 0, 1, 0, XX, tt, dd, e, verbose=.false.)
      call check_int(e, 1, 'missing directory is rejected')
   end subroutine test_missing_dir

   subroutine test_nonuniform()
      real(dp) :: tt(4), dd
      integer :: e
      tt = [0.0_dp, 0.1_dp, 0.2_dp, 0.35_dp]
      call uniform_dt(tt, dd, e)
      call check_int(e, 12, 'non-uniform sampling is rejected')
      tt = [0.0_dp, 0.1_dp, 0.2_dp, 0.3_dp]
      call uniform_dt(tt, dd, e)
      call check_int(e, 0, 'uniform sampling is accepted')
      call check_close(dd, 0.1_dp, 1.0e-14_dp, 'dt computed correctly')
      tt = [0.0_dp, 0.1_dp, 0.05_dp, 0.3_dp]
      call uniform_dt(tt, dd, e)
      call check_int(e, 11, 'non-increasing times are rejected')
   end subroutine test_nonuniform

end program test_end_to_end
